src/lib/mongoBackend/mongoSubCache.cpp

```diff
@@ -481,19 +441,73 @@ void mongoSubCacheDestroy(void)
 
 /* ****************************************************************************
 *
-* subCacheItemUpdate - 
+* tenantMatch - 
 */
-static void subCacheItemUpdate(const char* tenant, BSONObj* subP)
+static bool tenantMatch(const char* tenant1, const char* tenant2)
+{
+  if ((tenant1 == NULL) && (tenant2 == NULL))
+  {
+    return true;
+  }
+
+  if ((tenant1 == NULL) || (tenant2 == NULL))
+  {
+    return false;
+  }
+
+  if (strcmp(tenant1, tenant2) == 0)
+  {
+    return true;
+  }
+
+  return false;
+}
+
+
+
+/* ****************************************************************************
+*
+* mongoSubCacheItemLookup - 
+*
+* FIXME P7: lookups would be A LOT faster if the subCache used a hash-table instead of
+*           just a single linked link for the subscriptions.
+*           The hash could be adding each char in 'tenant' and 'subscriptionId' to a 8-bit
+*           integer and then we'd have a vector of 256 linked lists.
+*/
+CachedSubscription* mongoSubCacheItemLookup(const char* tenant, const char* subscriptionId)
+{
+  CachedSubscription* cSubP = mongoSubCache.head;
+
+  while (cSubP != NULL)
+  {
+    if ((tenantMatch(tenant, cSubP->tenant)) && (strcmp(subscriptionId, cSubP->subscriptionId) == 0))
+    {
+      return cSubP;
+    }
+
+    cSubP = cSubP->next;
+  }
+
+  return NULL;
+}
+
+
+
+/* ****************************************************************************
+*
+* mongoSubCacheItemInsert - 
+*/
+int mongoSubCacheItemInsert(const char* tenant, const BSONObj& sub)
 {
   //
   // 01. Check validity of subP parameter 
   //
-  BSONElement  idField = subP->getField("_id");
+  BSONElement  idField = sub.getField("_id");
 
   if (idField.eoo() == true)
   {
-    LM_E(("Database Error (error retrieving _id field in doc: '%s')", subP->toString().c_str()));
-    return;
+    LM_E(("Database Error (error retrieving _id field in doc: '%s')", sub.toString().c_str()));
+    return -1;
```

Comment:
Maybe a dictionary of codes that the fucntion may return should be added to the function comment preamble.